Improve: show ID number in Editor windows

SlySven opened the pull request:
#### Brief overview of PR changes/additions
This is to help the end-user identify and cross-reference the ID numbers used for permanent items in scripts AND the editor.

#### Motivation for adding to Mudlet
This PR should, perhaps, be considered alongside #6726 and #6742.

SlySven committed:
Improve: show ID number in Editor windows

This is to help the end-user identify and cross-reference the ID numbers
used for permanent items in scripts AND the editor.

This PR should, perhaps, be considered alongside #6726 and #6742.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Merge branch 'development' into Improve_showIDNumberInEditorWindows

SlySven committed:
Revise: remove dummy text used during form design

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

vadi2 committed:
Add: a placeholder text for aliases (#6759)

<!-- Keep the title short & concise so anyone non-technical can
understand it,
     the title appears in PTB changelogs -->
#### Brief overview of PR changes/additions
Add a placeholder text for aliases
#### Motivation for adding to Mudlet
I've seen a few incoming folk get confused by alias regexex, hopefully
this can steer them in the right direction


![image](https://user-images.githubusercontent.com/110988/230744979-742b9341-b421-4128-a697-5506a6debd5f.png)

#### Other info (issues closed, discussion etc)

SlySven committed:
Revise: fix ordering of widgets in aliases_main_area.ui form

They were out of order which messes with merging with other versions!

Signed-off-by: Stephen <slysven@virginmedia.com>

SlySven committed:
Merge branch 'development' into Improve_showIDNumberInEditorWindows

SlySven committed:
Merge branch 'development' into Improve_showIDNumberInEditorWindows

SlySven committed:
Merge branch 'development' into Improve_showIDNumberInEditorWindows

vadi2 reviewed (commented):
We haven't had much end user demand for this, but it does take up the users visual attention in the editor.

To balance out the low demand, how about we make it not so noticeable in the editor too? This way it'll be accessible to the minority who needs this, and not get in the way of all other folks.

SlySven committed:
Merge branch 'development' into Improve_showIDNumberInEditorWindows

SlySven committed:
Merge branch 'development' into Improve_showIDNumberInEditorWindows

SlySven commented:
> To balance out the low demand, how about we make it not so noticeable in the editor too? This way it'll be accessible to the minority who needs this, and not get in the way of all other folks.

How do you suggest that is done? What is it that makes it *so* noticeable? :confused:

vadi2 commented:
> What is it that makes it so noticeable?

I think the visible by default bit, and by far only a few people would be interested in looking at this - so not visible by default would be more appropriate. 

My idea would be to show the ID if you hover over the `Name:` label (and that label only, so not the text field that follows after).

SlySven commented:
> > What is it that makes it so noticeable?
> 
> I think the visible by default bit, and by far only a few people would be interested in looking at this - so not visible by default would be more appropriate.
> 
> My idea would be to show the ID if you hover over the `Name:` label (and that label only, so not the text field that follows after).

Well, hiding it in a tool-tip makes it less accessible to VI users - but a knob to show/hide them would be doable - the control would go in the "Editor" tab of the preferences, yes?

vadi2 commented:
Tooltip information is still accessible to VI users, can you confirm @ironcross32?

I'm also a bit lost on the motivation for all of the 3 related PRs - motivation for each one seems to point to each other... unclear what is this actually solving. Seems to be completing theoretically missing pieces sure (identifiers), but not something that we've had an issue with in practice :/

SlySven commented:
> I'm also a bit lost on the motivation for all of the 3 related PRs - motivation for each one seems to point to each other... unclear what is this actually solving. Seems to be completing theoretically missing pieces sure (identifiers), but not something that we've had an issue with in practice :/

They are all, I think, going to help when trying to track down issues and relationships between (generally permanent) things being manipulated in the Lua sub-system and relating them to what is seen in the editor. Particularly because we do not enforce uniqueness in names for many Mudlet items/groups and this can make it awkward to cross-reference them.

ironcross32 commented:
Tooltips are unaffected by this PR.

demonnic commented:
Have any of our users asked for this? Or is this a developer driven thing? How would we use this ID? What does it actually provide apart from a troubleshooting tool for some edge cases?

I'm still struggling to figure out the why, beyond "so it's visible"

SlySven commented:
> Tooltips are unaffected by this PR.

The question was more alone the lines of: "if we put the ID information in a tool-tip would it still be accessible to VI users?" At present - like you say - tooltips are unaffected by this PR.

SlySven commented:
> Have any of our users asked for this? Or is this a developer driven thing? How would we use this ID? What does it actually provide apart from a troubleshooting tool for some edge cases?
> 
> I'm still struggling to figure out the why, beyond "so it's visible"

I can't say that anyone has asked for this and yes it is a little "so it is visible"; but my original observation remains though: each Lua function that creates a permanent (as well as a temporary) item returns the ID number - and (IIRC) it also shows up in the Central Debug Console IIRC - but without having some way to see those IDs in the editor it seems as though we have missed something out.

SlySven committed:
Merge branch 'development' into Improve_showIDNumberInEditorWindows

SlySven committed:
Revise: add knob to show/hide the ID number display in the editor

This setting is also now saved on a per-profile basis.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven commented:
Well, I have added a knob on the "Editor" tab of the preferences (and removed the "Advanced" group box that just had the "Show Invisible Control Characters" option in it):

Before this PR: 
![image](https://github.com/Mudlet/Mudlet/assets/6163092/9114f378-6d24-4788-a090-a49a08b2b24c)

After this PR:
![image](https://github.com/Mudlet/Mudlet/assets/6163092/498435a1-0d05-4bdc-9003-38f0af7a054c)

SlySven committed:
Merge branch 'development' into Improve_showIDNumberInEditorWindows

vadi2 commented:
> and (IIRC) it also shows up in the Central Debug Console IIRC - but without having some way to see those IDs in the editor it seems as though we have missed something out.

Alright, yep, that seems logical.

Design-wise: what is the black border box around the ID label for?

SlySven commented:
> Design-wise: what is the black border box around the ID label for?

I think that is the default when you select `QFrame` as a container widget - I'll see if I can turn that off..

SlySven committed:
Revise: remove visible frame around text and ID number

Discovered during peer-review.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Fix: remove bogus clear() calls and insert missing setText(...) one

The last one meant that the ID numbers for Script items were not being
shown!

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

vadi2 reviewed (approved):
Looks good to me now!

Lets give others 2-3 days a chance to weigh before merging.

demonnic reviewed (approved):
looks ok. I still think the better solution is not to show IDs for perm triggers in the central debug console if the main point is to make them align, as the ID of a perm trigger can change between profile loads. For instance, if you make a bunch of temp triggers and then create a new one in the editor the ID will be one higher than the highest tempTrigger at that time, but when you reload the profile it will be in line with the other permanent items loaded from the profile.

I kind of prefer the separation of numeric ID for temp items and string name for perm ones, but since it's not displayed by default it at least won't cause a wave of people asking what they are and if they're important.

vadi2 commented:
Are we going in the right direction with this or is the real fix to remove them from showing up in the return of the function+debug console?

demonnic commented:
> Are we going in the right direction with this or is the real fix to remove them from showing up in the return of the function+debug console?

I would tend to think the latter. For perm items I don't really feel like the ID is helpful knowledge or even knowledge someone would look for except for us giving it to them in those messages. The divide of IDs for temp items and names for perm items is pretty well established at this point I think.

But I'm also not so opposed to exposing them that I'd block the PR now that it's not going to generate a bunch of extra questions

SlySven commented:
> ... I still think the better solution is not to show IDs for perm triggers in the central debug console if the main point is to make them align, as the ID of a perm trigger can change between profile loads. ...

Well we have to wrap the Lua script that each item has (well apart from Scripts as they are already Lua scripts) *something* and that is "Trigger" (or "Alias" or "Key" or etc. etc.) + ID-number + "()" + all the stuff in the Lua script box + "end" - so it is unavoidable that, say, "Trigger42" gets mentioned as that is the function that contains the stuff put in the box for the Trigger with ID number 42:

It is fair enough that the ID numbers are not necessarily going to be constant ***between*** sessions, but that they will be within any single session is still of use when debugging.

> I kind of prefer the separation of numeric ID for temp items and string name for perm ones, but since it's not displayed by default it at least won't cause a wave of people asking what they are and if they're important.

That separation is not as strong as you might think - and it is actually a bit confusing in places - e.g. the `permTimer(...)` function says:
> **Returns**
>
>   a unique id number for that timer.

whereas the C++ code for the `disableTimer(...)` function expects a **string** as an identifying argument and processes it as a string, NOT a number, all the way through the `(bool) TimerUnit::disableTimer(const QString& name)` even though it works for temporary timers:
```lua
lua myTempTimer=tempTimer(5, [[echo("Hello world!\n")]], true)
Hello world!
Hello world!
Hello world!
lua myTempTimer
14
Hello world!
Hello world!
Hello world!
lua disableTimer(myTempTimer)
true
lua enableTimer(myTempTimer)
true
Hello world!
Hello world!
```

:dizzy_face:

demonnic commented:
None of that changes the fact that literally nobody is asking to be able to see the ID numbers and we're only doing it because we think we should for 'completeness' because it gets exposed in a couple places though? I'm not saying they don't serve a purpose, but I am saying in over a decade of using Mudlet I've never once thought "Huh, I wonder what the ID for this permanent timer is?" and I don't think many of our users who are less deep nerds about the Mudlet internals than I are going to suddenly wonder it either. 

Like I said though, since it's tucked away and not gonna cause a bunch of "what's this ID and do I need to care about it?" questions just for existing it's not a bad thing to be able to bring them up in the rare instance it's helpful. But you really shouldn't write any code around the IDs since, as pointed out, **they change**
